### ISSUE TITLE
whiteout files present in singularity but not in docker container

### ISSUE THREAD

**Issue by kaczmarj** [apptainer/singularity#884]
## Version of Singularity:

```shell
$ singularity --version
2.3.1-HEAD.ge214d4e
```

```shell
$ docker --version
Docker version 17.07.0-ce-rc3, build 665d244
```

### Expected behavior

Consistency in presence of whiteout files between Docker and Singularity.

### Actual behavior

Whiteout files (e.g. ,`.wh..wh..opq`) are present in the Singularity container (pulled from DockerHub) but not in the Docker container.

### Steps to reproduce behavior

**Singularity**

```shell
$ singularity pull docker://tensorflow/tensorflow:1.3.0-gpu-py3
$ singularity shell --containall --shell bash tensorflow-1.3.0-gpu-py3.img
(container)$ find / -name ".wh.*" 2>/dev/null | grep -v "Permission denied"
```

**Docker**

```shell
$ docker run --rm -it --entrypoint bash tensorflow/tensorflow:1.3.0-gpu-py3
(container)$ find / -name ".wh.*" 2>/dev/null | grep -v "Permission denied"
```

related to #571

**Comment by vsoch** [comment]
Did you try the singularity commands using sudo? It could be permissions on pull / run (as Docker has write permissions). I'm not familiar with whiteout files - could you talk a bit about how/why they exist? For the longest time I thought those were python wheels :)

**Comment by vsoch** [comment]
I just tried with sudo and they indeed persist, I wonder why this happens? Do you think it would be appropriate for singularity to do a sweep of a container, always, and remove?

```
find / -name ".wh.*" -exec rm -f {} \;
```

**Comment by kaczmarj** [comment]
I know very little about whiteout files, but they seem to be part of the overlay filesystem. I found these resources ([linux kernel docs](https://github.com/torvalds/linux/blob/6b1c776d3efbda31085b6a9f3bc7f774511fafd9/Documentation/filesystems/overlayfs.txt#L92-L107) and [opencontainers layer spec](https://github.com/opencontainers/image-spec/blob/master/layer.md#whiteouts)) useful.

Here is a quote from [opencontainers](https://github.com/opencontainers/image-spec/blob/master/layer.md#opaque-whiteout):
> An opaque whiteout entry is a file with the name .wh..wh..opq indicating that all siblings are hidden in the lower layer.

So I am not sure whether it is safe to remove these files. I will try to dig into the docker source to see what they do with the opaque whiteout files. For what it's worth, the directories that contain whiteout files in singularity are present when running with docker (ie the directories+files are not hidden).

**Comment by vsoch** [comment]
ok, thanks for looking into this! It's definitely a big issue for some of our users.

**Comment by chrisgorgo** [comment]
I just run into this with 2.4. Looking forward to the fix!

**Comment by vsoch** [comment]
@chrisfilo see the PR I linked - if you have some bandwidth please give it a try and report back! Travis is having network issues for pulling docker images, but the local testing is far more important for this change.

**Comment by chrisgorgo** [comment]
Has this issue been resolved by #1181?

**Comment by dtrudg** [comment]
Closing this as PRs fixing whiteout issues have been merged to development. Let us know if there are any remaining problems. Thanks!
